crates/marketplace/src/testing/integration.rs

```diff
@@ -0,0 +1,213 @@
+
+        let leaf = Leaf::genesis(&validated_state, &instance_state).await;
+
+        let builder_state = BuilderState::<Types>::new(
+            ParentBlockReferences {
+                view_number: Types::Time::genesis(),
+                vid_commitment,
+                leaf_commit: leaf
+                    .commit(&UpgradeLock::<Types, TestVersions>::new())
+                    .await,
+                builder_commitment,
+            },
+            &receivers,
+            req_receiver,
+            Vec::new(), /* tx_queue */
+            Arc::clone(&global_state),
+            Duration::from_millis(1),
+            10,
+            Arc::new(instance_state),
+            Duration::from_secs(60),
+            Arc::new(validated_state),
+        );
+
+        builder_state.event_loop();
+
+        let hooks = Arc::new(NoHooks(PhantomData));
+
+        // create the proxy global state it will server the builder apis
+        let app = ProxyGlobalState::new(
+            global_state.clone(),
+            Arc::clone(&hooks),
+            builder_key_pair,
+            Duration::from_secs(60),
+        )
+        .into_app()
+        .expect("Failed to create builder tide-disco app");
+
+        let url_clone = url.clone();
+        async_spawn(async move {
+            tracing::error!("Starting builder app on {url_clone}");
+            if let Err(e) = app.serve(url_clone, StaticVersion::<0, 1> {}).await {
+                tracing::error!(?e, "Builder API App exited with error");
+            } else {
+                tracing::error!("Builder API App exited");
+            }
+        });
+
+        Box::new(MarketplaceBuilderTask {
+            hooks: config.hooks,
+            senders,
+        })
+    }
+}
+
+struct MarketplaceBuilderTask<Types>
+where
+    Types: NodeType,
+{
+    hooks: Arc<Box<dyn BuilderHooks<Types>>>,
+    senders: BroadcastSenders<Types>,
+}
+
+impl<Types> BuilderTask<Types> for MarketplaceBuilderTask<Types>
+where
+    Types: NodeType<Time = ViewNumber>,
+{
+    fn start(
+        self: Box<Self>,
+        stream: Box<
+            dyn futures::prelude::Stream<Item = hotshot::types::Event<Types>>
+                + std::marker::Unpin
+                + Send
+                + 'static,
+        >,
+    ) {
+        async_spawn(run_builder_service(self.hooks, self.senders, stream));
+    }
+}
+
+use hotshot_example_types::node_types::{Libp2pImpl, MemoryImpl, PushCdnImpl, TestTypes};
+use hotshot_testing::{
+    completion_task::{CompletionTaskDescription, TimeBasedCompletionTaskDescription},
+    test_builder::TestDescription,
+};
+
+cross_tests!(
+    TestName: test_success,
+    Impls: [MemoryImpl, Libp2pImpl, PushCdnImpl],
+    BuilderImpls: [MarketplaceBuilderImpl],
+    Types: [TestTypes],
+    Versions: [MarketplaceTestVersions],
+    Ignore: true,
+    Metadata: {
+        let mut metadata = TestDescription::default();
+        metadata.overall_safety_properties.transaction_threshold = 1;
```

Comment:
you could also add (either in addition to or instead of this):

` metadata.validate_transactions = hotshot_testing::test_builder::nonempty_block_threshold((80,100))`

This checks that we have nonempty blocks in e.g. 80% of leaves at the end of the test, but doesn't require it in every view

Reply:
Great, thanks! This sounds like a more reliable mechanism